### PR TITLE
feat(gitlab): refresh gitlab tokens

### PR DESCRIPTION
**Pull request by nickoferrall**
Fixes https://github.com/ParabolInc/parabol/issues/6592

### Testing

- Integrate with GitLab
- In gitlabLoaders.ts, change `if (!isValidAuth) {` to `if (true) {` so that we run the refresh logic every time the dataloader is called
- Comment out the upsert team member integration line: `upsertTeamMemberIntegrationAuth(newGitlabAuth),`
- Now, we're refreshing the token, but we're not upserting & storing the new access token. When refreshing the token, we invalidate the previous accessToken, so we should no longer have access to GitLab.
- Refresh the page and create a new Poker meeting. See that the GitLab issues don't show up in the Scope phase

https://www.loom.com/share/19d6b24adda649eb9175347856b97170

### CODE REVIEW

**Review by igorlesnenko — CHANGES_REQUESTED** [review]
I would suggest to store TTL in a `teamMemberIntegrationAuths` where token lives.

I don't see that redis gives us functionality that couldn't be implemented in postgres in this particular case.

If we store it in a PostgreSQL, we don't need any additional query, as we can just fetch datetime and check that in code.

As alternative, we can just use try catch, if gitlab returns an error we refresh the token.

What do you think?

File: packages/server/dataloader/gitlabLoaders.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+              sendToSentry(new Error('No refresh token in gitlabAuth'), {userId})
+              return null
+            }
+            const provider = await parent.get('integrationProviders').loadNonNull(providerId)
+            const {clientId, clientSecret, serverBaseUrl} = provider
+            const manager = new GitLabOAuth2Manager(clientId!, clientSecret!, serverBaseUrl!)
+            const oauthRes = await manager.refresh(refreshToken)
+            if (oauthRes instanceof Error) return null
+            const {accessToken, refreshToken: newRefreshToken, expires_in} = oauthRes
+            const expiresAtTimestamp = new Date().getTime() + (expires_in - 30) * 1000
```
Comment (nickoferrall) [inline packages/server/dataloader/gitlabLoaders.ts:37]: `expires_in is in seconds. Minus 30 to be safe. *1000 to convert it to milliseconds

**Comment by nickoferrall** [comment]
I've pushed an update using the postgres table rather than Redis. Ready for your review @igorlesnenko. 

Fingers crossed it can be squeezed into the ship this week so we can launch GitLab on Wednesday 🤞

File: packages/server/dataloader/gitlabLoaders.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+            const newGitlabAuth = {
+              ...gitlabAuth,
+              accessToken,
+              refreshToken: newRefreshToken,
+              expiresAt
+            }
+            upsertTeamMemberIntegrationAuth(newGitlabAuth)
+            return newGitlabAuth
+          }
+          return gitlabAuth as IGetTeamMemberIntegrationAuthQueryResult | null
```
Comment (Dschoordsch) [inline packages/server/dataloader/gitlabLoaders.ts:48]: -1 why add `| null` when we know it is `!== null` from l22?

File: packages/server/integrations/helpers/authorizeOAuth2.ts [hunk 3]
```diff
@@ -32,7 +32,12 @@ const transformBody = (contentType: string, body?: Record<string, string>): stri
 }
 
 export const authorizeOAuth2 = async <
-  TSuccess = {accessToken: string; refreshToken: string | undefined; scopes: string | undefined}
+  TSuccess = {
+    accessToken: string
+    refreshToken: string | undefined
+    scopes: string | undefined
+    expires_in: number | undefined
```
Comment (Dschoordsch) [inline packages/server/integrations/helpers/authorizeOAuth2.ts:39]: -1 we're fixing the other fields names to camelCase, so please do it also for `expiresIn`

**Review by Dschoordsch — CHANGES_REQUESTED** [review]
Only small things, feel free to merge after fixing. Did not test.

**Pull request merged**